### PR TITLE
Add implementation of typical sampling

### PR DESCRIPTION
**Pull request by cimeister**
# What does this PR do?

Adds an implementation of typical sampling (https://arxiv.org/abs/2202.00666) to the `generate` function

### CODE REVIEW

File: src/transformers/generation_utils.py [hunk 1]
```diff
@@ -2817,5 +2822,7 @@ def top_k_top_p_filtering(
 
     if 0 <= top_p <= 1.0:
         logits = TopPLogitsWarper(top_p=top_p, min_tokens_to_keep=min_tokens_to_keep)(None, logits)
-
+    
+    if 0 <= typical_p <= 1.0:
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:2826]: I think we can leave it out here

**Comment by patrickvonplaten** [comment]
Looks cool! Could you add one test? :-)

**Comment by cimeister** [comment]
> Looks cool! Could you add one test? :-)

Added a test and fixed the style issues :)

**Review by patrickvonplaten — APPROVED** [review]
Cool! Looks good for merge to me :-) 

Since @Narsil is on holiday - could you take a look here @patil-suraj ?

**Review by ngoquanghuy99 — APPROVED** [review]
This is awesome! Thanks for working on this.
Think @thomwolf should take a look :face_in_clouds:

**Comment by cimeister** [comment]
Of course! Are we waiting on @thomwolf at this point? Or is there something else that has to happen before merging?

File: tests/test_generation_logits_process.py [hunk 2]
```diff
@@ -191,6 +192,51 @@ def test_top_p_dist_warper(self):
         # first batch should keep three tokens, second batch would keep only 1, but due to `min_tokens_to_keep=2` keeps 2.
         self.assertListEqual((filtered_dist != 0.0).to(torch.long).sum(dim=-1).tolist(), [3, 2])
 
+    def test_typical_dist_warper(self):
```
Comment (patrickvonplaten) [inline tests/test_generation_logits_process.py:195]: Very nice test!

**Review by patrickvonplaten — APPROVED** [review]
Merging - very clean implementation. Thanks @cimeister

**Comment by thomwolf** [comment]
I confirm, very clean! Good job @cimeister

**Comment by LSinev** [comment]
@cimeister You are using `nansum` here, which is available in torch>=1.7.0. @patrickvonplaten Does approving this for main branch means `transformers` drop support for torch<1.7.0 ?

**Comment by cimeister** [comment]
Ah nice catch @LSinev 
@patrickvonplaten should we change to something like:
```
plogp = p * normalized 
ent = - plogp[~torch.isnan(plogp)].sum(-1, keepdim=True)
```

**Comment by patrickvonplaten** [comment]
Same answer here is given by @sgugger here: https://github.com/huggingface/transformers/pull/13292#discussion_r837529073 . Would be great if someone could open a PR to fix it :-)

**Pull request merged**